### ISSUE TITLE
[Question] One devcontainer multiple projects

### ISSUE THREAD

**Issue by fvelcker** [devcontainers/cli#414]
Hi,

First sorry if my question is obvious already answered... But it seems I can't find what I want to do. To keep it simple I detail our use case:

Basically we have several PHP projects. Some devs are on VSCode, some not.
What I try to do, is to have a prebuilt container not tight to any specific project. The prebuilt container is already built in a Gitlab pipeline.
I want to run that container on its own and each project can be attached to on developer demand (at least for vscode peeps).

I tried to run the container with the CLI, but it asked for a workspace (which I don't have specifically, so I put a random one) and when using the "attach to container" option, the workspace is not mounted (I followed this https://code.visualstudio.com/docs/devcontainers/attach-container#_attached-container-configuration-files).
If I run it directly with a docker run command, it doesn't stay up.

It is funny, when I write this it seems just obvious, but it seems I can't make it work! So if you have any input on what I am doing wrongly or I misunderstand anything, I'll be happy.

Thanks all

**Comment by jglogan** [comment]
Can you provide an example `devcontainer.json` and devcontainer CLI command line (and docker command line) that illustrates the problem that you are seeing?

It sounds like you a building an image using Gitlab CI and not a container.  Is it really that you want to have an exactly one container that you can use with multiple projects?  Or one container image that you can use to create multiple dev containers, each of which allows you to do development of a discrete project?

**Comment by fvelcker** [comment]
oops yes sorry mistyped, I prebuild the image not the container.

And yes, I want one running container for multiple projects (like we have 5 or 6 PHP projects, they are setup the same way, we don't need 6 times the same, especially we are using the container to only have the QA tools and IDE extensions).

For the `.devcontainer.json` , I used the one from the doc to keep it simple first.
```
{
  // Default path to open when attaching to a new container.
  "workspaceFolder": "/worskspaces/project1",

   // An array port numbers to forward
  "forwardPorts": [8000],

  // Container user VS Code should use when connecting
  "remoteUser": "vscode",
}
```

But I think I misunderstand. Looks when attaching to a container the `.devcontainer.json` doesn't really matter as it seems to be copied on the container at the first attach, to edit it it needs to be edited on the container. I thought first by attaching to the container, it would mount the project folder to the running container.
At the end I might just need to mount all projects folder on the container, and attach vscode to it, then navigate between project on the container.

**Comment by jglogan** [comment]
Do you want to bind mount your code from your host into the container so that your edits are available in both environments?

Or is it acceptable to have your code only a Docker volume that is mounted into the container?

You can have multiple projects in a container either way.  With bind mounts, you can bind mount a parent folder if all your projects are conveniently located in a single parent, or you can individually bind mount each project.  Conventionally, you'd mount things such that they all live under a /workspaces directory in the container.

With a named volume mount, you could have a post command that clones whatever projects you need under /workspace if they don't already exist there.

If you value build performance, the named volume should perform better than bind mounts on Mac (and WSL I believe but I don't have any experience there).

**Comment by chrmarti** [comment]
You can use `devcontainer build` to prebuild an image with a first devcontainer.json (or just `docker build` if you don't need any devcontainer specifics) and then push that image to a public or private registry from where you can use it in a second devcontainer.json that you would add to each project.

**Comment by chrmarti** [comment]
There is more information about pre-building images: https://code.visualstudio.com/docs/devcontainers/containers#_prebuilding-dev-container-images

**Comment by worldofgeese** [comment]
@chrmarti can you give an example of a good workflow for this? I'd prefer to have just one `devcontainer.json` file for building from and for my end users to use:

Preferably:
```json
{
  "name": "Devbox Remote Container",
  "image": "docker.io/worldofgeese/lbob-devcontainer:0.1.0",
  "build": {
    "dockerfile": "./Dockerfile",
    "context": "..",
    "cacheFrom": "docker.io/worldofgeese/lbob-devcontainer:0.1.0"
  },
```

where the Dev Container uses the image and `devcontainer build` follows the `build` key. Currently I need to either maintain two `devcontainer.json` files (where and what naming scheme? I'm monorepo) or delete the `build` key and add the `image` key after building and before pushing.

**Comment by fvelcker** [comment]
Hey,

Thanks for the answers.
I don't have any issues or questions with the prebuilt image, that's all good (at least for that thread :)).

I was only wondering how to use 1 container for multiple projects that use the same language. Like if the project folder could be mounted on the fly while attaching it to the container. 
But yes, as @jglogan said looks like for my case I'll have to mount the folders one by one (or the parent folder, but I would have to group them per language) into the container.

**Comment by chrmarti** [comment]
@worldofgeese For using two devcontainer.json files: You could add a `prebuild-devcontainer` (or any other name) folder to your monorepo containing a `.devcontainer.json` or `.devcontainer/devcontainer.json` and use that folder as the workspace folder when building your base image.

@fvelcker Unfortunately the mount points can't be changed after the container was created. (This is a limitation from Docker, not specific to dev containers.)

**Comment by fvelcker** [comment]
@chrmarti yeah, I learnt that thanks to that thread. It also made reassess what I am doing.
Anyway, thank you and thanks @jglogan for your help.
